elastic/kibana
# [Screenshotting] Move PDF generation code to screenshotting

## Summary

To create PDFs we use the `PdfMaker`. This happens as a post-processing step in Reporting once we have received PNGs from the screenshotting plugin.

This change moves the PDF generation logic into screenshotting, so that Reporting does not contain the implementation for creating PDFs. This change brings the following benefits:

* Screenshotting logic is encapsulated in one place for all screenshot file formats
* We don't need to worry about sharing `Layout` outside of Screenshotting
* If we move Screenshotting to a different server, the functionality for creating a PDF will move with it

## Design

I've decided to keep `observable.ts` the same as it is today. There is already quite a lot of complexity contained in the module for generating PNG screenshots. Instead, screenshot results are piped directly into formatting which is treated as a new concept in the screenshotting plugin that handles layout and file type and result "shaping".

This resulted in the following:

```
./server
├── browsers
├── config
├── formats <-- new directory
├── index.ts
├── layouts
├── mock.ts
├── plugin.test.ts
├── plugin.ts
├── screenshots
└── utils.ts

5 directories, 5 files
```

In this way, we can keep all of the tests and functionality for `screenshots` the same, and compose formatting functionality instead of increasing the responsibility of `screenshots`.

One last consideration was the combination of format and layouts. Formats now encapsulate file type and layouts. PNG format accepts only one type of layout: `preserve_layout`. This is because it does not currently support anything else, and likely will not need to. PDFs support `preserve_layout`, `print` and `canvas`. At the moment there is no runtime enforcement of these but I think we could open a follow-up issue to address this.

## How to test

1. Start Kibana, load sample data and turn on trial license
2. Generate a PDF, PDF print-optimised and PNG report on Dashboard
3. Generate a PDF and PDF without margins report from Canvas
4. Generate a PNG report from visualize
5. Do a sanity check on the results of all of the above generated reports

## Notes

* It looks like there is bug that we were are not passing through `renderErrors` at the moment: https://github.com/elastic/kibana/commit/976de5182e361251205e5af0c966764316519a04

### Checklist

This is purely a refactor, functionality is expected to remain the same. Any changes to tests will be due to changing APIs.
